File: Kalorie/Kalorie/Core/UseCases/UpdateMealTypeTimesUseCase.swift
//
//  UpdateMealTypeTimesUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 30.06.2026.
//

import Foundation

protocol UpdateMealTypeTimesUseCaseProtocol {
    func callAsFunction(_ mealTypes: [MealTypeDomain]) async throws
}

struct UpdateMealTypeTimesUseCase: UpdateMealTypeTimesUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(_ mealTypes: [MealTypeDomain]) async throws {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let dtos = mealTypes.map { mealType in
            (
                item: MealTypeDTO(
                    id: mealType.id,
                    name: mealType.name,
                    startTime: mealType.startTime.timeIntervalSince1970,
                    endTime: mealType.endTime.timeIntervalSince1970
                ),
                id: "\(mealType.id)"
            )
        }
        try await dataProvider.batchSetAsync(dtos, in: Constants.Firestore.mealTypes(userId: userId))
    }
}

struct UpdateMealTypeTimesUseCaseFake: UpdateMealTypeTimesUseCaseProtocol {

    // MARK: - Functions

    func callAsFunction(_ mealTypes: [MealTypeDomain]) async throws {}
}

File: Android/app/src/main/java/antoni/kalorie/core/usecases/UpdateMealTypeTimesUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.MealTypeDTO
import antoni.kalorie.core.networking.batchSetAsync
import antoni.kalorie.core.utils.Constants

interface UpdateMealTypeTimesUseCaseProtocol {
    suspend operator fun invoke(mealTypes: List<MealTypeDomain>)
}

class UpdateMealTypeTimesUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : UpdateMealTypeTimesUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(mealTypes: List<MealTypeDomain>) {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val dtos = mealTypes.map { mealType ->
            MealTypeDTO(
                id = mealType.id,
                name = mealType.name,
                startMinutes = mealType.startMinutes,
                endMinutes = mealType.endMinutes,
            ) to mealType.id
        }
        dataProvider.batchSetAsync(dtos, inCollection = Constants.Firestore.mealTypes(userId))
    }
}
